internal/service/admin/adminservice/signup.go

```diff
@@ -51,3 +63,81 @@ func (s *adminService) Signup(ctx context.Context, u *ent.Admin) error {
 
 	return nil
 }
+
+/*
+Angus: Here's an example of how we can use the Decorator pattern to move noisy tracing and logging out of your
```

Comment:
Wow, that absolutely insane how much cleaner and more efficient this is. I'm definitely going to implement this
in my template. I take a bit of time to understand this chain but that very perfect. Yes it's very easy to test and
to understand. Thanks a lot for this insane example! 🤯

I also have a question about folder structure, do you think that we need to decouple the service in differente files or keep all in one file? I think that it's better to decouple in different files because with a very big service it
can be difficult to find the good function, but maybe it's better to keep all in one file to have a better overview of the service. What do you think about that?

Reply:
I definitely advise you to practice writing your own HTTP server (`http.Handler`) and client (`http.RoundTripper) middleware, and chaining them together. Bonus points if you can figure out how to unit test them without spawning an HTTP server! I found these chains confusing to begin with too, but it's an extremely useful pattern you can apply in many situations.

How you structure your files is quite personal, but I can tell you what I prefer. To start with, I'd keep everything belonging to one domain in the same package, including the decorators, but put your decorator definitions in a file called `middleware.go`.

If your application gets really big, with many decorators and several domains, you could consider pulling all decorators from across domains into a separate package, such as `internal/domain/middleware`. This happens very rarely in practice though.